pkg/collector/corechecks/gpu/nvml.go

```diff
@@ -0,0 +1,84 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+//go:build linux
+
+// Package gpu defines the agent corecheck for
+// the GPU integration
+package gpu
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/NVIDIA/go-nvml/pkg/nvml"
+)
+
+var initOnce sync.Once
+
+type gpuDevice struct {
+	nvml.Device
+}
+
+func wrapNvmlError(ret nvml.Return) error {
+	if ret == nvml.SUCCESS {
+		return nil
+	}
+
+	return errors.New(nvml.ErrorString(ret))
+}
+
+func ensureNvmlInit() error {
+	var err error
+	initOnce.Do(func() {
+		err = wrapNvmlError(nvml.Init())
+	})
+
+	return err
+}
+
+func getGPUDevices() ([]gpuDevice, error) {
+	err := ensureNvmlInit()
+	if err != nil {
+		return nil, err
+	}
+
+	count, ret := nvml.DeviceGetCount()
+	if err := wrapNvmlError(ret); err != nil {
+		return nil, fmt.Errorf("cannot get number of GPU devices: %w", err)
+	}
+
+	var devices []gpuDevice
+
+	for i := 0; i < count; i++ {
+		device, ret := nvml.DeviceGetHandleByIndex(i)
+		if err := wrapNvmlError(ret); err != nil {
+			return nil, fmt.Errorf("cannot get handle for GPU device %d: %w", i, err)
+		}
+
+		devices = append(devices, gpuDevice{device})
+	}
+
+	return devices, nil
+}
+
+func (d *gpuDevice) GetNumMultiprocessors() (int, error) {
+	devProps, ret := d.GetAttributes()
+	if err := wrapNvmlError(ret); err != nil {
+		return 0, fmt.Errorf("cannot get device attributes: %w", err)
+	}
+
+	return int(devProps.MultiprocessorCount), nil
+}
+
+func (d *gpuDevice) GetMaxThreads() (int, error) {
```

Comment:
I double-checked just in case, this function does refer to the number of cores that can execute threads. Naming is a bit confusing as a GPU has different multiprocessors and each one has a number of cores, but I think this is the correct API call.

Reply:
so should we rename it to `GPUCoresCount` or `GetNumGPUCores` (same as the underlying function of the go-nvml), and why do we need this wrapper at all and not just use the go-nvml function directly?

Reply:
At first it contained a calculation, but even with that calculation removed I prefer having the wrapper so that it's explicit this is the way to get the maximum number of threads. It also makes it easier if for some reason we need to have alternative logic to calculate the number of threads in the future.

Reply:
ok, but the API returns the number of Cores not the number of threads, correct? to get the number of threads, we must multiple it by some other parameters of the device (you actually have a nice [diagram](https://datadoghq.atlassian.net/wiki/spaces/EBPFTEAM/pages/4084236609/The+CUDA+software+and+hardware+architecture) that explains it)

Reply:
That was the idea I had at first, but I discovered it was the wrong approach, specially because in the documentation it's not clearly specified which numbers are maximums depending on the CUDA capability version and which ones depend on the actual hardware. The correct calculation is the number of multiprocessors multiplied by the number of cores per multiprocessor, this function gives us that product computed already (in the Tesla T4 case, for example, it's 2560 cores, which is 40 multiprocessors multiplied by 64 cores for each multiprocessor).

Reply:
Also another source of confusion I had: the difference "resident" and "active". For example, the [tech specs](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#features-and-technical-specifications-technical-specifications-per-compute-capability) say that the maximum number of resident threads for multiprocessor is 1024, which doesn't mean that it can execute 1024 threads at the same time, just that it can have 1024 threads assigned/queued in a single multiprocessor.